### ISSUE TITLE
[CI] UpdateMappingOnClusterTests.testTimestampMergingConflicts

### ISSUE THREAD

**Issue by brwe** [elastic/elasticsearch#10297]
http://build-us-00.elastic.co/job/es_core_master_window-2008/1351/consoleFull

reproducible with
```
 mvn clean test -Dtests.seed=18576849DC7E894A -Dtests.class=org.elasticsearch.index.mapper.update.UpdateMappingOnClusterTests -Dtests.method="testTimestampMergingConflicts" -Des.logger.level=DEBUG -Des.node.mode=local -Dtests.assertion.disabled=org.elasticsearch -Dtests.security.manager=true -Dtests.nightly=false -Dtests.client.ratio=0.0 -Dtests.heap.size=512m -Dtests.jvm.argline="-server -XX:+UseSerialGC -XX:-UseCompressedOops -XX:+AggressiveOpts -Djava.net.preferIPv4Stack=true" -Dtests.locale=ar_BH -Dtests.timezone=Pacific/Nauru -Dtests.processors=4

```

on master.

`doc_values: true` is lost for _timestamp. Probably related to https://github.com/elastic/elasticsearch/pull/10209 although not sure if the commit caused this or just exposed an existing bug.

**Comment by rjernst** [comment]
Here is what happens:

1. The randomized default template is setup with `_timestamp` having `doc_values: false` explicitly.  However, this is serialized *without* any `doc_values` setting because `false` is the default value for `_timestamp`.
2. The first mapping is added on index creation. It contains `fielddata.format: doc_values`. Since the default template was serialized without any doc values settings, this takes precedence and doc values is enabled, and part of the mappings serialization.
3. The mappings update is made. This has `fieldata.format: array`.  This causes `_timestamp` to use the default for doc values, which is `false`. Again, since this is the default, it is skipped during serialization.
4. The test fails when comparing the serialized mappings from (2) and (3), the first having `doc_values: true` and the second having no doc values setting.

The fix is to change how the doc values setting is stored in `AbstractFieldMapper`.  Instead of using a `boolean`, we instead need `Boolean`, so that a `null` values means the default was used, and a non-null value indicates we should *always* write out the `doc_values` setting (it was explicitly set).

**Comment by rjernst** [comment]
Note that this is an edge case because there are 2 ways to specify doc values.  If there were just one, skipping serialization of the default value would be fine (there would be nothing else to potentially supersede it).

**Comment by brwe** [comment]
Actually, it seems to me this is some sort of race condition. If I add a sleep of 1s here: https://github.com/elastic/elasticsearch/blob/master/src/test/java/org/elasticsearch/index/mapper/update/UpdateMappingOnClusterTests.java#L230 before the first mapping is requested, then the test passes. 

In addition, the mapping that changes the fielddata format to `fieldata.format: array` is never applied (point 3. in the description) because there is a mapping conflict for this update and so all mappings should be untouched by the update. `doc_values: true` should remain too. If something changed then I would assume this is a bug. Also, is it even possible to switch from `doc_values` from `true` to `false`? I thought not?

**Comment by rjernst** [comment]
I agree it looks like there is a race condition.  However, I think the race condition is only there because with the bug I fix in #10302, the mappings are serialized differently than when they are read.  So any refresh-mapping action would cause the parsed mappings to change.

While I don't know what triggers refresh-mapping actions, when I add your 1 second sleep with a printout, I notice this:
```
2015-03-30 11:00:58,899][DEBUG][org.elasticsearch.cluster.service] [node_s0] processing [refresh-mapping [index][[type]]]: execute
[2015-03-30 11:00:58,899][INFO ][org.elasticsearch.index.mapper.update] --> Sleeping for 1 sec
...
[2015-03-30 11:00:58,923][DEBUG][org.elasticsearch.cluster.service] [node_s0] cluster state updated, version [10], source [refresh-mapping [index][[type]]]
```

So, a refresh-mapping is kicked off just before the sleep, and the sleep gives it about enough time to finish (so that by the time the get mapping is executed, it has the refreshed view, which now has `doc_values: true`.  Also, the log message from the failure is actually reversed.  The "got" is what we got before, and the "expected" is the second retrieval at the end of the test.  The `assertThat` call has the arguments in the wrong order, which I think has led to part of the confusion. At first when the index template is applied, the in memory version has `doc_values: false`.  Once the refresh-mapping action is complete, we have `doc_values: true` because the explicit doc values setting has been lost.
